src/libraries/Microsoft.Extensions.DependencyInjection/src/ServiceLookup/ServiceDescriptorCacheKey.cs

```diff
@@ -0,0 +1,65 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
+using System.Diagnostics.CodeAnalysis;
+
+namespace Microsoft.Extensions.DependencyInjection.ServiceLookup
+{
+    internal readonly struct ServiceIdentifier : IEquatable<ServiceIdentifier>
+    {
+        public object? ServiceKey { get; }
+
+        public Type ServiceType { get; }
+
+        public ServiceIdentifier(Type serviceType)
+        {
+            ServiceType = serviceType;
+        }
+
+        public ServiceIdentifier(object? serviceKey, Type serviceType)
+        {
+            ServiceKey = serviceKey;
+            ServiceType = serviceType;
+        }
+
+        public static ServiceIdentifier FromDescriptor(ServiceDescriptor serviceDescriptor)
+            => new ServiceIdentifier(serviceDescriptor.ServiceKey, serviceDescriptor.ServiceType);
+
+        public static ServiceIdentifier FromServiceType(Type type) => new ServiceIdentifier(null, type);
+
+        public bool Equals(ServiceIdentifier other)
+        {
+            if (ServiceKey == null && other.ServiceKey == null)
+            {
+                return ServiceType.Equals(other.ServiceType);
+            }
+            else if (ServiceKey != null && other.ServiceKey != null)
+            {
+                return ServiceType.Equals(other.ServiceType) && ServiceKey.Equals(other.ServiceKey);
+            }
+            return false;
+        }
+
+        public override bool Equals([NotNullWhen(true)] object? obj)
+        {
+            return obj is ServiceIdentifier && Equals((ServiceIdentifier)obj);
+        }
+
+        public override int GetHashCode()
+        {
+            unchecked
+            {
+                if (ServiceKey == null)
+                {
+                    return ServiceType.GetHashCode();
+                }
+                return ((ServiceType?.GetHashCode() ?? 23) * 397) ^ ServiceKey.GetHashCode();
```

Comment:
Nit:

```suggestion
                return ((ServiceType.GetHashCode() ?? 23) * 397) ^ ServiceKey.GetHashCode();
```

Can we not use `HashCode.Combine` because of the nestandard2.0 target? `HashCode.Combine` doesn't have an unchecked block. Do we really need it here?

Reply:
Yep, it's because of netstandard2.0. If you have an alternative I am interested.

Reply:
Not sure if it is feasible or worthwhile here, but in RDG, we use shared source to access the method. ([ref](https://source.dot.net/#Microsoft.AspNetCore.Http.RequestDelegateGenerator/src/Shared/HashCode.cs)).